(feat): Add a minimum polling interval for CHANGE_ON_EVENT subscription type

kostinkirill opened the pull request:
Improvement for https://github.com/apache/plc4x/issues/2335

Added minimum polling interval for CHANGE_ON_EVENT subscription type

kostinkirill committed:
(feat): Add a minimum polling interval for CHANGE_ON_EVENT subscription type

chrisdutz reviewed (changes requested):
Generally ok ... but in this case I guess we would need matching Duration-counterparts for:

- addChangeOfStateTagAddress (Without consumer)
- addChangeOfStateTag (Without consumer)
- addChangeOfStateTag (With consumer)

kostinkirill committed:
(feat): added method with DURATION:
addChangeOfStateTagAddress (Without consumer)
addChangeOfStateTagAddress (With consumer)
addChangeOfStateTag (Without consumer)

kostinkirill commented:
@chrisdutz added methods with dueration: 

- addChangeOfStateTagAddress (Without consumer)
- addChangeOfStateTagAddress (With consumer)
- addChangeOfStateTag (Without consumer)

chrisdutz commented:
I assume the next step would be to extend the ADS and possibly the OPC-UA driver to support this, right? Are you planning on continuing to work on this? I's prefer to mentor you through this as it's a bit like the "give a man a fish and he's fed for a day, teach him to fish and he's fed for life" sort of thing ;-)

chrisdutz commented:
The question now is: If a driver doesn't implement this Duration-based limiting ... should we throw an "unsupported" error, or should we silently forward it to the duration-less version and log a message?

kostinkirill commented:
> I assume the next step would be to extend the ADS and possibly the OPC-UA driver to support this, right? Are you planning on continuing to work on this? I's prefer to mentor you through this as it's a bit like the "give a man a fish and he's fed for a day, teach him to fish and he's fed for life" sort of thing ;-)

I believe ADS drive already supports it, I've already checked the behavior and `minInterval` is applied for CHANGE_ON_EVENT subscriptions and works as epxected. 
This the code where the `minInterval` is set up as `cycleTimeInMs` in `AdsAddDeviceNotificationRequest`
`AdsProtocolLogic#executeSubscribe:
```
return new AmsTCPPacket(new AdsAddDeviceNotificationRequest(configuration.getTargetAmsNetId(), configuration.getTargetAmsPort(),
                    configuration.getSourceAmsNetId(), configuration.getSourceAmsPort(),
                    0, getInvokeId(),
                    directAdsTag.getIndexGroup(),
                    directAdsTag.getIndexOffset(),
                    adsDataTypeTableEntry.getSize() * numberOfElements,
                    tag.getPlcSubscriptionType() == PlcSubscriptionType.CYCLIC ? AdsTransMode.CYCLIC : AdsTransMode.ON_CHANGE, // if it's not cyclic, it's on change or event
                    0, // there is no api for that yet
                    tag.getDuration().orElse(Duration.ZERO).toMillis()));
```
(the last argument)

As for OPCUA, I haven't checked it, but based on the code it should work as well

kostinkirill commented:
> The question now is: If a driver doesn't implement this Duration-based limiting ... should we throw an "unsupported" error, or should we silently forward it to the duration-less version and log a message?

I'd say drive implements it

chrisdutz commented:
Interesting ... wasn't expecting that to work ;-) 

But yeah ... having a look at the code, it actually makes sense. Cool ... so I guess this is ready to being merged ;-)

Thanks :-)
